hazelcast/src/main/java/com/hazelcast/sql/impl/expression/CaseExpression.java

```diff
@@ -0,0 +1,156 @@
+            whenExpressions[index] = (Expression<Boolean>) operands[i];
+            thenExpressions[index] = operands[i + 1];
+            index += 1;
+        }
+        return new CaseExpression<>(whenExpressions, thenExpressions, operands[operands.length - 1], resultType);
+    }
+
+    @Override
+    public int getFactoryId() {
+        return SqlDataSerializerHook.F_ID;
+    }
+
+    @Override
+    public int getClassId() {
+        return SqlDataSerializerHook.EXPRESSION_CASE;
+    }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
+        out.writeInt(whenExpressions.length);
+
+        for (int i = 0; i < whenExpressions.length; i++) {
+            out.writeObject(whenExpressions[i]);
+            out.writeObject(thenExpressions[i]);
+        }
+
+        out.writeObject(elseExpression);
+        out.writeObject(resultType);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void readData(ObjectDataInput in) throws IOException {
+        int branchesSize = in.readInt();
+
+        whenExpressions = new Expression[branchesSize];
+        thenExpressions = new Expression[branchesSize];
+
+        for (int i = 0; i < branchesSize; i++) {
+            whenExpressions[i] = in.readObject();
+            thenExpressions[i] = in.readObject();
+        }
+
+        elseExpression = in.readObject();
+        resultType = in.readObject();
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public T eval(Row row, ExpressionEvalContext context) {
+        for (int i = 0; i < whenExpressions.length; i++) {
+            Expression<Boolean> condition = whenExpressions[i];
+
+            Boolean conditionHolds = condition.eval(row, context);
+            if (TernaryLogic.isTrue(conditionHolds)) {
+                return (T) thenExpressions[i].eval(row, context);
+            }
+        }
+
+        return (T) elseExpression.eval(row, context);
+    }
+
+    @Override
+    public QueryDataType getType() {
+        return resultType;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        CaseExpression that = (CaseExpression) o;
+        return Arrays.equals(whenExpressions, that.whenExpressions)
+                && Arrays.equals(thenExpressions, that.thenExpressions)
+                && Objects.equals(elseExpression, that.elseExpression)
+                && Objects.equals(resultType, that.resultType);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(Arrays.hashCode(whenExpressions), Arrays.hashCode(thenExpressions), elseExpression, resultType);
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder builder = new StringBuilder("CaseExpression{");
+        int len = whenExpressions.length;
+        for (int i = 0; i < len; i++) {
```

Comment:
We typically do not do line breaks or so in the `toString`. The proper implementation is simply to dump the existing fields one by one, without any sophisticated processing. You may refer to other operators for reference.